cli/deno_dir.rs

```diff
@@ -421,102 +418,91 @@ impl DenoDir {
-
-                if maybe_initial_module_url.is_none() {
-                  maybe_initial_module_url = Some(module_url);
-                }
-
-                // Not yet completed. Follow the redirect and loop.
-                Ok(Loop::Continue((
-                  dir,
-                  maybe_initial_module_url,
-                  new_module_url,
-                )))
-              }
-              FetchOnceResult::Code(source, maybe_content_type) => {
-                // TODO: move caching logic outside this function
-                // We land on the code.
-                dir.save_source_code_headers(
-                  &module_url,
-                  maybe_content_type.clone(),
-                  None,
-                ).unwrap();
-
-
-                dir.save_source_code(
-                  &module_url,
-                  &source
-                ).unwrap();
-
-                if let Some(redirect_source_url) = &maybe_initial_module_url {
-                  dir.save_source_code_headers(
-                    redirect_source_url,
-                    maybe_content_type.clone(),
-                    Some(module_url.to_string())
-                  ).unwrap()
-                }
-
-                let filepath = dir
-                  .deps_cache
-                  .location
-                  .join(dir.deps_cache.get_cache_filename(&module_url));
-
-                let media_type = map_content_type(
-                  &filepath,
-                  maybe_content_type.as_ref().map(String::as_str),
-                );
-
-                let source_file = SourceFile {
-                  url: module_url,
-                  redirect_source_url: maybe_initial_module_url,
-                  filename: filepath,
-                  media_type,
-                  source_code: source.as_bytes().to_owned(),
-                };
-
-                Ok(Loop::Break(Some(source_file)))
-              }
-            }
-          },
-        )
+    let module_uri = url_into_uri(&module_url);
+
+    let dir = self.clone();
+    let module_url = module_url.clone();
+
+    // Single pass fetch, either yields code or yields redirect.
+    let f = http_util::fetch_string_once(module_uri).and_then(
+      move |fetch_once_result| {
+        match fetch_once_result {
+          FetchOnceResult::Redirect(uri) => {
+            // If redirects, update module_name and filename for next looped call.
+            let new_module_url = Url::parse(&uri.to_string()).expect("http::uri::Uri should be parseable as Url");
+
+            dir.save_source_code_headers(
+              &module_url,
+              None,
+              Some(new_module_url.to_string())
+            ).unwrap();
+
+            // Explicit drop to keep reference alive until future completes.
+            drop(download_job);
+
+            let module_url = module_url.clone();
+
+            // Recurse
```

Comment:
Is there a mechanism to avoid infinite recursion?

Reply:
I added basic one just now